v2_explainer.md

```diff
@@ -0,0 +1,76 @@
+# Picture-in-Picture V2 Explained
+
+## Background: Picture-in-Picture V1
+
+Picture-in-Picture launched for HTMLVideoElement early in 2019. The API was
+designed to be extensible, like the Fullscreen API but at the moment is only
+defined for video elements. The implementation takes the video layer and moves
+it into a window that follows display rules specific to Picture-in-Picture
+(always on top, fixed aspect ratio, etc.).
+
+## What's Missing from V1?
+
+Various partners expressed frustration around the limitations of the API. The
+feature requests fell into two categories: customising the look & feel and
+customising the user experience. The former can technically be done using canvas
+elements but the latter requires a different model that would require user
+interaction and therefore no longer be compatible with Android, iOS and macOS
+APIs.
+
+## Goals
+
+  1. **Allow sites to Picture-in-Picture arbitary content**
+  2. **A site should be able to choose whether that content is interactive.** If
+the platform (e.g. Android) does not support interactive content then the site
+should be able to easily fallback to non-interactive content
+
+## Non-Goals
+
+The Picture-in-Picture window is not a popup. Therefore, the content should not
+live beyond the lifetime of the page.
+
+## Sample Usage
+
+Here's how a page might request Picture-in-Picture for it's media controls and
+then move the video element to the new window. When the window is closed it will
+then move the video element back to the page.
+
+```html
+<video id=video>
+<div id=controls>
+```
+
+```js
+const controls = document.getElementById('controls');
+const video = document.getElementById('video');
+
+const pipWindow = await controls.requestPictureInPicture({
+  aspectRatio: 1,
+  interactable: true
+});
+
+pipWindow.document.body.appendChild(video);
+
+pipWindow.addEventListener('leavepictureinpicture', () => {
+  document.body.appendChild(video);
+});
+```
+
+The `aspectRatio` will be used to determine the size of the Picture-in-Picture
+window.
+
+## Alternatives Considered
+
+We considered using window.open with a additional `picture-in-picture` and
+`interactive` options. However, exposing that `interactive` or
+`picture-in-picture` is not supported in `window.open()` could only return
+`null` which is not very descriptive. We could change this behaviour but it
```

Comment:
`window.open` already throws errors that could be used to detect whether or not Picture-in-Picture is not possible. See https://html.spec.whatwg.org/multipage/dynamic-markup-insertion.html#document-open-steps.

Reply:
These steps are for `document.open` whereas `window.open` can only return the window or null: https://html.spec.whatwg.org/multipage/window-object.html#window-open-steps. We would have to add support for throwing or returning an exception which sites may not expect and may cause breakage with normal `window.open` calls.

Reply:
You're right. I was looking at the wrong algorithm.
The correct one you pointed out (https://html.spec.whatwg.org/multipage/window-object.html#window-open-steps) can also raise errors. See impl in Chrome: https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/frame/local_dom_window.cc?type=cs&sq=package:chromium&g=0&l=1468

Reply:
Ah yes I see. I have added clarification.